src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/EnumeratingWhitelist.java

```diff
@@ -50,69 +54,127 @@
 
     protected abstract List<FieldSignature> staticFieldSignatures();
 
-    // TODO should precompute hash sets of signatures, assuming we document that the signatures may not change over the lifetime of the whitelist (or pass them in the constructor)
+    ConcurrentHashMap<String, Boolean> permittedCache = new ConcurrentHashMap<String, Boolean>();  // Not private to facilitate testing
+
+    private void cacheSignatureList(List<Signature> ...sigs) {
+        for (List<Signature> list : sigs) {
+            for (Signature s : list) {
+                if (!s.isWildcard()) { // Cache entries for wildcard signatures will never be accessed and just waste space
+                    permittedCache.put(s.toString(), Boolean.TRUE);
+                }
+            }
+        }
+    }
+
+    /** Prepopulates the "permitted" cache. */
+    public void precache() {
```

Comment:
Should be `final`. Or better, not `public` at all, since this does not look like something you should ever call from outside. At least, it neglects to call `permittedCache.clear()` before it starts, so it would leave behind stale cache entries if you were to modify your implementation and then call `precache` yourself.

Whatever the behavior, this needs to be specified in class Javadoc, since IIUC we are changing the expected semantics of this class: you may no longer dynamically return different values from `xxxSignatures` methods.